R/pmcmc_nested_parameters.R

```diff
@@ -0,0 +1,488 @@
+        parameters$names)
+
+      private$proposal_kernel_fixed <- proposals$kernel_fixed
+      private$proposal_kernel_varied <- proposals$kernel_varied
+      private$proposal_fixed <- proposals$proposal_fixed
+      private$proposal_varied <- proposals$proposal_varied
+      names(private$proposal_kernel_varied) <- parameters$pop
+      names(private$proposal_varied) <- parameters$pop
+
+      # transform function
+      if (is.null(transform)) {
+        private$transform <- as.list
+      } else {
+        private$transform <- assert_is(transform, "function")
+      }
+
+      invisible(self)
+    },
+
+    ##' @description Return the names of the parameters
+    names = function() {
+      unlist(private$param_names, use.names = FALSE)
+    },
+
+    ##' @description Return the names of the populations
+    populations = function() {
+      names(private$varied_parameters)
+    },
+
+        ##' @description Return a `data.frame` with information about
+    ##' parameters (name, min, max, discrete, fixed or varied).
+    ##'
+    ##' @param population For parameter sets including varied parameters,
+    ##' `population` specifies which population to summarise. If `NULL` then
+    ##'  returns summary of each population as a list.
+    summary = function(population = NULL) {
+      summarise_population <- function(population) {
+        if (!(population %in% self$populations())) {
+          stop(sprintf("Expected 'population' in %s.",
+            str_collapse(self$populations())))
+        }
+        type <- rep("fixed", length(self$names()))
+        type[self$names() %in% private$param_names$varied] <- "varied"
+        min <- c(r6_private(private$fixed_parameters)$min,
+          r6_private(private$varied_parameters[[population]])$min)
+        max <- c(r6_private(private$fixed_parameters)$max,
+          r6_private(private$varied_parameters[[population]])$max)
+        discrete <- c(r6_private(private$fixed_parameters)$discrete,
+          r6_private(private$varied_parameters[[population]])$discrete)
+
+        data_frame(
+          name = self$names(),
+          min = min,
+          max = max,
+          discrete = discrete,
+          type = type
+        )
+      }
+
+      if (is.null(population)) {
+        sum <- lapply(self$populations(), summarise_population)
+        names(sum) <- self$populations()
+        sum
+      } else {
+        summarise_population(population)
+      }
+    },
+
+    ##' @description Return the initial parameter values as a named numeric
+    ##' vector or named matrix for multiple populations.
+    initial = function() {
+
+      varied <- vapply(private$varied_parameters, function(x) x$initial(),
+        numeric(length(private$param_names$varied)))
+
+      # edge case - 1 varied
+      if (length(private$param_names$varied) == 1) {
+        varied <- t(as.matrix(varied))
+        rownames(varied) <- private$param_names$varied
+      }
+
+      if (is.null(private$fixed_parameters)) {
```

Comment:
`length(private$fixed_parameters) == 0` might capture intent better than `is.null`

Reply:
Really? It is stored as NULL if not present

Reply:
I'd store it as `list()` if not present (i.e., none are present)